# **CASE STUDY 1**

> click **> Run All** on the toolbar above to get all the results at once

# 1. Overview

**1.1 - Dataset Context**

Dataset Name : **employee transaction log**

This dataset is a simulation for audit practice purposes, generated with the help of generative AI. This audit is conducted to identify intentional anomalies from the Dataset.

> *Preparation : To get started, run the code below (shortcut : ctrl + Enter)*

In [ ]:
import pandas as pd
import numpy as np

# The following is the employee transaction log data
data = {
    "employee_id": ["E001", "E002", "E003", "E004", "E005",
                    "E003", "E006", "E007", "E008", "E009"],
    "name": ["Budi Santoso", "Siti Rahma", "Andi Wijaya", "Dewi Kusuma",
              "Reza Pratama", "Andi Wijaya", "Maya Indah",
              "Fajar Nugroho", "Linda Sari", "Hendra Gunawan"],
    "department": ["Finance", "HR", "Finance", "IT", "Finance",
                   "Finance", "Marketing", "", "HR", "IT"],
    "amount": [1500000, 750000, 15000000, None, 1200000,
               9800000, 500000, 650000, -200000, 900000],
    "date": ["2024-01-05", "2024-01-06", "2024-01-07", "2024-01-08",
             "2024-01-09", "2024-01-10", "2024-01-11", "2024-01-12",
             "2024-01-13", "2024-01-14"],
    "status": ["approved", "approved", "approved", "pending",
                "approved", "approved", "approved", None,
                "approved", "rejected"]
}

df = pd.DataFrame(data)

# Output
print(df)

  employee_id            name department      amount        date    status
0        E001    Budi Santoso    Finance   1500000.0  2024-01-05  approved
1        E002      Siti Rahma         HR    750000.0  2024-01-06  approved
2        E003     Andi Wijaya    Finance  15000000.0  2024-01-07  approved
3        E004     Dewi Kusuma         IT         NaN  2024-01-08   pending
4        E005    Reza Pratama    Finance   1200000.0  2024-01-09  approved
5        E003     Andi Wijaya    Finance   9800000.0  2024-01-10  approved
6        E006      Maya Indah  Marketing    500000.0  2024-01-11  approved
7        E007   Fajar Nugroho               650000.0  2024-01-12      None
8        E008      Linda Sari         HR   -200000.0  2024-01-13  approved
9        E009  Hendra Gunawan         IT    900000.0  2024-01-14  rejected


**1.2 - Dataset Structure**

In [ ]:
# Method to check dataset structure
print(f"Dataset header preview")
print('='*50)
print(df.head(1))
print("\n")

print(f"Dataset shape (rows, columns)")
print('='*50)
print(df.shape)
print("\n")

print(f"Dataset column types")
print('='*50)
print(df.dtypes)

Dataset header preview
  employee_id          name department     amount        date    status
0        E001  Budi Santoso    Finance  1500000.0  2024-01-05  approved


Dataset shape (rows, columns)
(10, 6)


Dataset column types
employee_id     object
name            object
department      object
amount         float64
date            object
status          object
dtype: object


From the output above, the Dataset Structure is:
- **6** Headers; **employee_id, name, department, amount, date, status**
- **10** Rows, **6** Columns
- Object data type; **employee_id, name, department, date, status**
- Float64 data type (decimal); **amount**
- Dataset type; **object**

> Dataset has been identified and audit can be carried out.

**1.3 - Statistical Summary**

In [ ]:
# Method to view statistical summary of the dataset
print(df.describe())

             amount
count  9.000000e+00
mean   3.344444e+06
std    5.316804e+06
min   -2.000000e+05
25%    6.500000e+05
50%    9.000000e+05
75%    1.500000e+06
max    1.500000e+07


**Initial findings** from the dataset statistics:

- Minimum value (min) is negative **-200,000**; indicates a potential anomaly
- Outlier (from max) **15,000,000**; the range is too far from the mean **3,344,444**
- Count of '**amount**' is only **9**, a difference of **1** from the total rows of **10**

# 2. Analysis

**2.1 - Audit Checklist**
1. Which transactions have an amount above the overall average?
2. How many rows contain missing values (NaN), and in which columns?
3. Which employee appears more than once and is their total transaction amount reasonable?

**2.2 - Checklist 1 Solution**

Which transactions have an amount above the overall average?

**Checklist 1 - Step 1** : Finding the average transaction value

In [ ]:
# Method to calculate the mean transaction value
df_mean_transaction = df['amount'].mean()

# Output
print(f"Average transaction value : {df_mean_transaction}")

Average transaction value : 3344444.4444444445


**Checklist 1 - Step 2** : Finding transactions where 'amount' exceeds the average

In [ ]:
# Method to filter amount values greater than the mean
df_above_average = df[df['amount'] > df_mean_transaction]

# Output
print(df_above_average)

  employee_id         name department      amount        date    status
2        E003  Andi Wijaya    Finance  15000000.0  2024-01-07  approved
5        E003  Andi Wijaya    Finance   9800000.0  2024-01-10  approved


**Checklist 1 - Conclusion**

Found **2 transactions** exceeding the average value:

1. index 2; &nbsp;&nbsp;&nbsp; E003 &nbsp; Andi Wijaya &nbsp; Finance &nbsp; 15,000,000.0 &nbsp; 2024-01-07 &nbsp; approved
2. index 5; &nbsp;&nbsp;&nbsp; E003 &nbsp; Andi Wijaya &nbsp; Finance &nbsp;&nbsp; 9,800,000.0 &nbsp; 2024-01-10 &nbsp; approved

---

**2.3 - Checklist 2 Solution**

How many rows contain missing values (NaN), and in which columns?

**Checklist 2 - Step 1** : Finding missing values in the dataset

In [ ]:
# Method to convert empty strings to NaN
df['department'] = df['department'].replace("", np.nan)

# Method to find missing values
df_missing_values = df.isnull().sum()

# Output
print(df_missing_values)

employee_id    0
name           0
department     1
amount         1
date           0
status         1
dtype: int64


**Checklist 2 - Conclusion**

Missing values found: **1** in **amount** column, **1** in **department** column, and **1** in **status** column

---

**2.4 - Checklist 3 Solution**

Which employee appears more than once and is their total transaction amount reasonable?

**Checklist 3 - Step 1** : Finding employees who appear more than once

In [ ]:
# Method to group data by namployee_id and count occurrences
df_employee_frequency = df.groupby('employee_id')['name'].value_counts()

# Output
print(df_employee_frequency)

employee_id  name          
E001         Budi Santoso      1
E002         Siti Rahma        1
E003         Andi Wijaya       2
E004         Dewi Kusuma       1
E005         Reza Pratama      1
E006         Maya Indah        1
E007         Fajar Nugroho     1
E008         Linda Sari        1
E009         Hendra Gunawan    1
Name: count, dtype: int64


**Checklist 3 - Step 2** : Finding employees with unreasonable transaction amounts

In [ ]:
# Method to group by employee_id and sum the total amount
df_total_per_employee = df.groupby('employee_id').agg({'name': 'first', 'amount': 'sum'})

# Output
print(df_total_per_employee)

                       name      amount
employee_id                            
E001           Budi Santoso   1500000.0
E002             Siti Rahma    750000.0
E003            Andi Wijaya  24800000.0
E004            Dewi Kusuma         0.0
E005           Reza Pratama   1200000.0
E006             Maya Indah    500000.0
E007          Fajar Nugroho    650000.0
E008             Linda Sari   -200000.0
E009         Hendra Gunawan    900000.0


**Checklist 3 - Conclusion**

Found employee **E003** name **Andi Wijaya** appearing **2 times** in the transaction log

---

# 3. Audit Report

**3.1 - Findings**

1. Employee **Andi Wijaya** made **2 transactions** (**15,000,000** and **9,800,000**) with a combined total of **24,800,000** that exceeding the average of **3,344,444** per transaction. There is also an indication of **Splitting Transaction** to avoid a transaction threshold. This anomaly requires further investigation as it directly impacts transaction value. Risk : **High**

2. Employee **Dewi Kusuma** has an invalid transaction value **(NaN)**. Needs to be investigated, It might a data entry error. Risk : **High**

3. Employee **Linda Sari** has a **negative** transaction value of **-200,000**. Is this legitimate/valid? Risk : **Medium**

4. Employee **Fajar Nugroho** has no value in the **department** column and a **None** status. At risk of being silently ignored by the system. Risk : **Medium**

**3.2 - Recommendations**

1. Investigate and verify the anomalous transactions by Andi Wijaya. This can be followed up by the Finance/HR department.
2. Define input validation rules for columns that should contain numeric values. Directed to the Developer/Engineer department for improvement.
3. The validity of negative values needs to be determined. Policy required by Finance/HR.
4. Define standards for empty columns and status values outside accepted categories. Values outside categories should be standardized, for example, replacing blank values with "None". Developer/Engineer needs to make the adjustment.

**3.3 - Limitations**

This dataset contains only 10 rows of data and is generated with the help of generative AI for audit practice purposes; transaction bias findings need to be verified with a larger dataset before any conclusion can be drawn.